Documentation/design/installconfig.md

```diff
@@ -0,0 +1,183 @@
+    VPCID           string   `json:"tectonic_aws_external_vpc_id,omitempty" yaml:"vpcID,omitempty"`
+    WorkerSubnetIDs []string `json:"tectonic_aws_external_worker_subnet_ids,omitempty" yaml:"workerSubnetIDs,omitempty"`
+}
+
+// Master converts master related config.
+type Master struct {
+    CustomSubnets    map[string]string `json:"tectonic_aws_master_custom_subnets,omitempty" yaml:"customSubnets,omitempty"`
+    EC2Type          string            `json:"tectonic_aws_master_ec2_type,omitempty" yaml:"ec2Type,omitempty"`
+    ExtraSGIDs       []string          `json:"tectonic_aws_master_extra_sg_ids,omitempty" yaml:"extraSGIDs,omitempty"`
+    IAMRoleName      string            `json:"tectonic_aws_master_iam_role_name,omitempty" yaml:"iamRoleName,omitempty"`
+    MasterRootVolume `json:",inline" yaml:"rootVolume,omitempty"`
+}
+
+// Worker converts worker related config.
+type Worker struct {
+    CustomSubnets    map[string]string `json:"tectonic_aws_worker_custom_subnets,omitempty" yaml:"customSubnets,omitempty"`
+    EC2Type          string            `json:"tectonic_aws_worker_ec2_type,omitempty" yaml:"ec2Type,omitempty"`
+    ExtraSGIDs       []string          `json:"tectonic_aws_worker_extra_sg_ids,omitempty" yaml:"extraSGIDs,omitempty"`
+    IAMRoleName      string            `json:"tectonic_aws_worker_iam_role_name,omitempty" yaml:"iamRoleName,omitempty"`
+    LoadBalancers    []string          `json:"tectonic_aws_worker_load_balancers,omitempty" yaml:"loadBalancers,omitempty"`
+    WorkerRootVolume `json:",inline" yaml:"rootVolume,omitempty"`
+}
+```
+
+#### libvirt specfic configuration
+
+```go
+type Libvirt struct {
+    URI           string `json:"tectonic_libvirt_uri,omitempty" yaml:"uri"`
+    SSHKey        string `json:"tectonic_libvirt_ssh_key,omitempty" yaml:"sshKey"`
+    QCOWImagePath string `json:"tectonic_coreos_qcow_path,omitempty" yaml:"imagePath"`
+    Network       `json:",inline" yaml:"network"`
+    MasterIPs     []string `json:"tectonic_libvirt_master_ips,omitempty" yaml:"masterIPs"`
+}
+
+// Network describes a libvirt network configuration.
+type Network struct {
+    Name      string `json:"tectonic_libvirt_network_name,omitempty" yaml:"name"`
+    IfName    string `json:"tectonic_libvirt_network_if,omitempty" yaml:"ifName"`
+    DNSServer string `json:"tectonic_libvirt_resolver,omitempty" yaml:"dnsServer"`
+    IPRange   string `json:"tectonic_libvirt_ip_range,omitempty" yaml:"ipRange"`
+}
+```
+
+### 4.0 Installer InstallConfig
+
+```go
+type InstallConfig struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+
+    ClusterID       uuid.UUID `json:"clusterID"`
+    SSHKey          string `json:"sshKey"`
+    Networking      `json:"networking"`
+    Machines        map[string]MachineSet `json:"machines"` // MachineSet https://github.com/kubernetes-sigs/cluster-api/blob/dfa371814c0b39a284662b897ba9b5df4749e973/pkg/apis/cluster/v1alpha1/machineset_types.go#L40
```

Comment:
This seems the wrong abstraction point (mentioned elsewhere).

Reply:
I really don't want to overdesign node groups here, but I also don't want to deviate so much that we don't leave room for it.

Reply:
@smarterclayton cluster-operator also has TODOs that indicate move to upstream MachineSet objects. That's why i thought using it here will be useful.